### PR TITLE
Boxplot stats w/ equal quartiles

### PR DESCRIPTION
**Pull request by phobson**
See #5331 

Addresses the concern raised in the issue above and cleans up the docstring.

Previous behavior available through the `autorange` kwarg.

### CODE REVIEW

File: lib/matplotlib/cbook.py [hunk 1]
```diff
@@ -1867,39 +1867,41 @@ def delete_masked_points(*args):
...
-    '''
-    Returns list of dictionaries of staticists to be use to draw a series of
-    box and whisker plots. See the `Returns` section below to the required
-    keys of the dictionary. Users can skip this function and pass a user-
-    defined set of dictionaries to the new `axes.bxp` method instead of
-    relying on MPL to do the calcs.
+def boxplot_stats(X, whis=1.5, autorange=False, bootstrap=None,
+                  labels=None):
+    """
+    Returns list of dictionaries of staticists to be use to draw a
```
Comment (mdboom) [inline lib/matplotlib/cbook.py]: sp. statistics

"to be use" -> "to use"

File: lib/matplotlib/cbook.py [hunk 2]
```diff
@@ -1867,39 +1867,41 @@ def delete_masked_points(*args):
...
+    series of box and whisker plots. See the `Returns` section below to
+    the required keys of the dictionary. Users can skip this function
+    and pass a user-defined set of dictionaries to the new `axes.bxp`
+    method instead of relying on MPL to do the calcs.
```
Comment (mdboom) [inline lib/matplotlib/cbook.py]: calcs -> calculations

File: lib/matplotlib/cbook.py [hunk 3]
```diff
@@ -1867,39 +1867,41 @@ def delete_masked_points(*args):
...
-    labels : sequence
-        Labels for each dataset. Length must be compatible with dimensions
-        of `X`
+        As a float, determines the reach of the whiskers past the first
+        and third quartiles (e.g., Q3 + whis*IQR, QR = interquartile
+        range, Q3-Q1). Beyond the whiskers, data are considered outliers
+        and are plotted as individual points. This can be set this to an
+        ascending sequence of percentile (e.g., [5, 95]) to set the
+        whiskers at specific percentiles of the data. Finally, `whis`
+        can be the string 'range' to force the whiskers to the min and
```
Comment (mdboom) [inline lib/matplotlib/cbook.py]:
```
``'range'``
```

Comment (mdboom) [inline lib/matplotlib/cbook.py]: min -> minimum
max -> maximum

File: lib/matplotlib/cbook.py [hunk 4]
```diff
@@ -1867,39 +1867,41 @@ def delete_masked_points(*args):
...
+        max of the data. In the edge case that the 25th and 75th
+        percentiles are equivalent, `whis` can be automatically set to
+        'range' via the ``autorange`` option.
```
Comment (mdboom) [inline lib/matplotlib/cbook.py]: Double ticks around 'range'

Single ticks around autorange

File: lib/matplotlib/cbook.py [hunk 5]
```diff
@@ -1867,39 +1867,41 @@ def delete_masked_points(*args):
...
+    autorange : bool (default = False)
```
Comment (mdboom) [inline lib/matplotlib/cbook.py]: bool, optional

File: lib/matplotlib/cbook.py [hunk 6]
```diff
@@ -1867,39 +1867,41 @@ def delete_masked_points(*args):
...
+        When True and the data are distributed such that the  25th and
```
Comment (mdboom) [inline lib/matplotlib/cbook.py]: "When True <comma>"

File: lib/matplotlib/cbook.py [hunk 7]
```diff
@@ -1867,39 +1867,41 @@ def delete_masked_points(*args):
...
+        75th percentiles are equal, ``whis`` is set to "range" such that
```
Comment (mdboom) [inline lib/matplotlib/cbook.py]:
```
``'range'``
```

File: lib/matplotlib/cbook.py [hunk 8]
```diff
@@ -1867,39 +1867,41 @@ def delete_masked_points(*args):
...
+        the whisker ends are at the min and max of the data.
```
Comment (mdboom) [inline lib/matplotlib/cbook.py]: min -> minimum
max -> maximum

**Comment by mdboom** [comment]
Cool.  Much improved.  :+1:

File: lib/matplotlib/cbook.py (hunk 4 above)
Comment (phobson) [inline lib/matplotlib/cbook.py]: so that's

```
def boxplots_stats(...):
""" ...
``'range'`` via the `autorange` option.
...
... """
```

?

Comment (mdboom) [inline lib/matplotlib/cbook.py]: Yes.

**Comment by phobson** [comment]
@mdboom thanks for the comments. I think I got them all as they came in :smile: 

...aaaaaand here's the part where I pitch the idea that we add in an option to pass your own bootstrapper.

Something like:

```python
def boxplots_stats(..., bootstrap_fxn=None):
    if bootstrap_fxn is None:
        bootstrap_fxn = _bootstrap_median

    # ...
    def _compute_conf_interval(data, med, iqr, bootstrap):
        if bootstrap is not None:
            # Do a bootstrap estimate of notch locations.
            # get conf. intervals around median
            CI = bootstrap_fxn(data, N=bootstrap)
            notch_min = CI[0]
            notch_max = CI[1]
        else:
            N = len(data)
            notch_min = med - 1.57 * iqr / np.sqrt(N)
            notch_max = med + 1.57 * iqr / np.sqrt(N)

    # yada yada 
```

**Comment by tacaswell** [comment]
I am starting to think that box plots are a complicated enough topic that they should be spun off into a sub-project (which is allowed to do things like require pandas and scipy).  Probably take violin plots with them too.

**Comment by QuLogic** [comment]
Isn't that essentially seaborn?

**Comment by WeatherGod** [comment]
We would also still want basic boxplot/violinplot functionality for those
who do not need to delve into pandas or seaborn.

**Comment by phobson** [comment]
rebased with current master (branch had gotten stale)

File: lib/matplotlib/axes/_axes.py [hunk 9]
```diff
@@ -3063,119 +3065,104 @@ def boxplot(self, x, notch=None, sym=None, vert=None, whis=None,
...
+            Labels for each dataset. Length must be compatible with
+            dimensions  of ``x``.
+        manage_xticks : bool, optional (True)
             If the function should adjust the xlim and xtick locations.
+        autorange : bool, optional (False)
+            When `True` and the data are distributed such that the  25th and
+            75th percentiles are equal, ``whis`` is set to ``'range'`` such
+            that the whisker ends are at the minimum and maximum of the
+            data.
+        meanline : bool, optional (False)
```
Comment (tacaswell) [inline lib/matplotlib/axes/_axes.py]: the `meanline` docs should move back up to be in the right order to match the signature.

Comment (tacaswell) [inline lib/matplotlib/axes/_axes.py]: eh, maybe not, I can go either way on this on further consideration.

**Comment by tacaswell** [comment]
tagged this as a bug fix, but I am not sure if that is the correct tag for this.

**Comment by phobson** [comment]
@tacaswell should I rebase on 1.5.X or 2?

**Comment by tacaswell** [comment]
on to master then we will back-port the merge to where ever we decide this will go.

**Comment by QuLogic** [comment]
It looks like you committed the conflicts in the SVG files:
```
.../result_images/test_axes/boxplot_autorange_false_whiskers-expected.svg:470: parser error : StartTag: invalid element name
<<<<<<< HEAD
    ^
```

**Comment by phobson** [comment]
gah -- thanks for the heads up. @QuLogic

**Comment by phobson** [comment]
Any objections to removing the PDF/SVG files for this test entirely?

**Comment by tacaswell** [comment]
No objection from me, makes the tests go faster :)

**Comment by phobson** [comment]
unrelated failure in py3.4
```
======================================================================
FAIL: matplotlib.tests.test_mathtext.test_mathfont_dejavuserif_02.test
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/matplotlib/matplotlib/venv/lib/python3.4/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/travis/build/matplotlib/matplotlib/lib/matplotlib/testing/decorators.py", line 55, in failer
    result = f(*args, **kwargs)
  File "/home/travis/build/matplotlib/matplotlib/lib/matplotlib/testing/decorators.py", line 259, in do_test
    '(RMS %(rms).3f)'%err)
matplotlib.testing.exceptions.ImageComparisonFailure: images not close: /home/travis/build/matplotlib/matplotlib/result_images/test_mathtext/mathfont_dejavuserif_02.png vs. /home/travis/build/matplotlib/matplotlib/result_images/test_mathtext/mathfont_dejavuserif_02-expected.png (RMS 29.168)
----------------------------------------------------------------------
```

**Comment by phobson** [comment]
... another rebase with current master

File: lib/matplotlib/axes/_axes.py [hunk 10]
```diff
@@ -3075,119 +3077,104 @@ def boxplot(self, x, notch=None, sym=None, vert=None, whis=None,
...
+            If `True` (and ``showmeans`` is `True`), will try to render
+            the mean as a line spanning the full width of the box
+            according to ``meanprops`` (see below). Not recommended if
+            ``shownotches`` is also True. Otherwise, means will be shown
+            as points.
+
+        Additional Options
+        ---------------------
+        The following boolean options toogle the drawing of individual
```
Comment (QuLogic) [inline lib/matplotlib/axes/_axes.py]: toogle -> toggle

File: lib/matplotlib/axes/_axes.py [hunk 11]
```diff
@@ -3075,119 +3077,104 @@ def boxplot(self, x, notch=None, sym=None, vert=None, whis=None,
...
+        components of the boxplots:
+            - showcaps: the caps on the ends of whiskers
+              (default is True)
+            - showbox: the central box (default is True)
+            - showfliers: the outlierd beyone the caps (default is True)
```
Comment (QuLogic) [inline lib/matplotlib/axes/_axes.py]: outlierd -> outliers
beyone -> beyond

File: lib/matplotlib/cbook.py [hunk 12]
```diff
@@ -1760,39 +1760,42 @@ def delete_masked_points(*args):
...
-    Returns list of dictionaries of staticists to be use to draw a series of
-    box and whisker plots. See the `Returns` section below to the required
-    keys of the dictionary. Users can skip this function and pass a user-
-    defined set of dictionaries to the new `axes.bxp` method instead of
-    relying on MPL to do the calcs.
+def boxplot_stats(X, whis=1.5, autorange=False, bootstrap=None,
+                  labels=None):
+    """
+    Returns list of dictionaries of statistics used to draw a series
+    of box and whisker plots. See the `Returns` section below to the
```
Comment (QuLogic) [inline lib/matplotlib/cbook.py]: to -> for (I think; the sentence is unclear.)

File: lib/matplotlib/axes/_axes.py [hunk 13]
```diff
@@ -3075,119 +3077,104 @@ def boxplot(self, x, notch=None, sym=None, vert=None, whis=None,
...
+            using a Gaussian-based asymptotic approximation (see McGill,
+            R., Tukey, J.W., and Larsen, W.A., 1978, and Kendall and
+            Stuart, 1967). Otherwise, bootstrap specifies the number of
+            times to bootstrap the median to determine its 95%
+            confidence intervals. Values between 1000 and 10000 are
+            recommended.
+        usermedians : array-like, optional
+            An array or sequence whose first dimension (or length) is
+            compatible with ``x``. This overrides the medians computed
+            by matplotlib for each element of *usermedians* that is not
```
Comment (QuLogic) [inline lib/matplotlib/axes/_axes.py]: Missed changing `usermedians` from asterisks to backticks (and on line below.)

File: lib/matplotlib/axes/_axes.py [hunk 14]
```diff
@@ -3075,119 +3077,104 @@ def boxplot(self, x, notch=None, sym=None, vert=None, whis=None,
...
+            `None`. When an element of *usermedians* == None, the median
```
Comment (QuLogic) [inline lib/matplotlib/axes/_axes.py]: `==` -> `is`

File: lib/matplotlib/axes/_axes.py [hunk 15]
```diff
@@ -3075,119 +3077,104 @@ def boxplot(self, x, notch=None, sym=None, vert=None, whis=None,
...
+            will be computed by matplotlib as normal.
+        conf_intervals : array-like, optional
+            Array or sequence whose first dimension (or length) is
+            compatible with ``x`` and whose second dimension is 2. When
+            the current element of ``conf_intervals`` is not None, the
```
Comment (QuLogic) [inline lib/matplotlib/axes/_axes.py]: What does "current" mean here? I'm guessing "corresponding to `x`" is the meaning intended.

**Comment by phobson** [comment]
Bump -- just gave this another rebase to keep it current with master

**Comment by phobson** [comment]
Build failures are unrelated. Something's wacky with colorbars, e.g.,
`FAIL: matplotlib.tests.test_colorbar.test_colorbar_closed_patch.test`

**Comment by WeatherGod** [comment]
Yes, this came about due to a merge of a different PR last night, I think.

**Comment by tacaswell** [comment]
Sorry, I broke all the branches.  Merged a PR that passed before we put the zero-tolerance in.  There were some regions of those images where the 8bit blue value changed by 1

**Comment by jenshnielsen** [comment]
The failure on appveyor is 

```
FAIL: Github issue #1256 identified a bug in Line.draw method 
---------------------------------------------------------------------- 
Traceback (most recent call last):
  File "C:\conda\envs\test-environment\lib\site-packages\nose\case.py", line 197, in runTest
    self.test(*self.arg)
  File "c:\projects\matplotlib\lib\matplotlib\testing\decorators.py", line 152, in wrapped_callable
    func(*args, **kwargs)
  File "c:\projects\matplotlib\lib\matplotlib\tests\test_lines.py", line 61, in test_invisible_Line_rendering
    assert_true(slowdown_factor < slowdown_threshold)
AssertionError: False is not true 
 ```

which is known to be flaky

**Comment by phobson** [comment]
Bump -- give me a shout if y'all want any changes made to this.

File: lib/matplotlib/cbook.py [hunk 16]
```diff
@@ -1760,39 +1760,42 @@ def delete_masked_points(*args):
...
 
 
-def boxplot_stats(X, whis=1.5, bootstrap=None, labels=None):
-    '''
-    Returns list of dictionaries of staticists to be use to draw a series of
-    box and whisker plots. See the `Returns` section below to the required
-    keys of the dictionary. Users can skip this function and pass a user-
-    defined set of dictionaries to the new `axes.bxp` method instead of
-    relying on MPL to do the calcs.
+def boxplot_stats(X, whis=1.5, autorange=False, bootstrap=None,
```
Comment (tacaswell) [inline lib/matplotlib/cbook.py]: Is this considered a public function?  If so the new arg needs to go last.

**Comment by tacaswell** [comment]
Read through this, other than my one comment :+1:.

The docstrings are much better.

Can this get a note in `api_changes`?   To check my understanding:

 - this is a minor API change (ex the plot will be different for same inputs) but
    1. only in semi-pathological edge case
    2. the change will make edge case more consistent with normal case
    3. the way it was is considered confusing by those skilled in the art of boxplots
    4. there is a way to get the old plots back

**Comment by phobson** [comment]
@tacaswell your understanding matches mine.

and to make sure I'm clear -- I don't modify `api_changes.rst`directly, I just add a short, single file: `api_changes/2016-03-12-PMH.rst`?

**Comment by tacaswell** [comment]
Yes.  The individual files helps prevent rebase-due-to-doc-conflicts
thrashing.

**Comment by jenshnielsen** [comment]
Backport wasn't clean (conflict in removed svg file) so doing it via #6153

**Comment by phobson** [comment]
thanks for the merge and help, everyone!

**Pull request merged**